cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java

```diff
@@ -543,6 +549,51 @@ public Boolean call() throws Exception {
     assertNoAccess(appId);
   }
 
+  @Test
+  public void testCrossNSDatasetAccessWithAuth() throws Exception {
+    createAuthNamespace();
+    getNamespaceAdmin().create(new NamespaceMeta.Builder()
+                                 .setName(CrossNsDatasetAccessApp.DATASET_OUTPUT_SPACE).build());
+    NamespaceId datasetOutputSpace = new NamespaceId(CrossNsDatasetAccessApp.DATASET_OUTPUT_SPACE);
+    addDatasetInstance(datasetOutputSpace.toId(), "keyValueTable", "store");
+    ApplicationManager appManager = deployApplication(AUTH_NAMESPACE.toId(), CrossNsDatasetAccessApp.class);
+
+    // Send stream and run the flowlet as BOB. Should fail
+    StreamManager streamManager = getStreamManager(AUTH_NAMESPACE.toId(), CrossNsDatasetAccessApp.STREAM_NAME);
+    for (int i = 0; i < 10; i++) {
+      streamManager.send(String.valueOf(i).getBytes());
+    }
+    Authorizer authorizer = getAuthorizer();
+    authorizer.grant(AUTH_NAMESPACE, BOB, ImmutableSet.of(Action.ALL));
+    SecurityRequestContext.setUserId(BOB.getName());
+    try {
+      FlowManager flowManager = appManager.getFlowManager(CrossNsDatasetAccessApp.FLOW_NAME);
+      flowManager.start();
+      Assert.fail("Bob don't have the privlege to write to the dataset. Failed");
+    } catch (Exception e) {
+      // Expected
+    }
+
+    // Switch back to ALICE. Should Success
+    SecurityRequestContext.setUserId(ALICE.getName());
+    FlowManager flowManager = appManager.getFlowManager(CrossNsDatasetAccessApp.FLOW_NAME);
+    flowManager.start();
+    for (int i = 0; i < 10; i++) {
+      streamManager.send(String.valueOf(i).getBytes());
+    }
+    flowManager.getFlowletMetrics("saver").waitForProcessed(10, 30, TimeUnit.SECONDS);
+
+    DataSetManager<KeyValueTable> dataSetManager = getDataset(datasetOutputSpace.toId(), "store");
+    KeyValueTable results = dataSetManager.get();
+    for (int i = 0; i < 10; i++) {
+      byte[] key = String.valueOf(i).getBytes(Charsets.UTF_8);
+      Assert.assertArrayEquals(key, results.read(key));
+    }
+
+    flowManager.stop();
+    getNamespaceAdmin().delete(datasetOutputSpace.toId());
```

Comment:
The auth test will delete the `AuthNamespace` in the end. So all the resources this test case created in this namespace will also be deleted right?